README
INTRODUCTION
------------

Another In-Place Editor is a jQuery plugin that turns any element or group of elements
into an Ajax in-place editor using only one line of code. It’s written using the jQuery
Library, which is freely available at http://jquery.com.


SUPPORT AND BUG REPORTS
-----------------------

Bug reports, as well as feature requests, may be submitted at:

http://code.google.com/p/jquery-in-place-editor/issues/list

Alternatively, you may email the developer directly: davehauenstein@gmail.com


QUICK START
-----------

include jquery and the editin place scripts in the head section of your html page

<script type="text/javascript" src="path/to/js/jquery.js"></script>
<script type="text/javascript" src="path/to/js/jquery.editinplace.packed.js"></script>

add the following code to the head of your html page

<script type="text/javascript" charset="utf-8">
    $(".inplace-editor").editInPlace({
        url: "path/to/server/script.php"
    });
</script>

create a div with the class 'inplace-editor'

<div class="inplace-editor"></div>

create a server side handler script (php, python, etc...)

script.php

the following parameters will be sent via a POST requeset to the server script

update_value
element_id
original_html

ex:
$_POST['update_value']
$_POST['element_id']
$_POST['original_html']


DOCUMENTATION
-------------

Extensive Documenataion can be found on the home page of this product: http://code.google.com/p/jquery-in-place-editor/



CHANGELOG
2010-05-14  Martin Haecker  <spamfaenger@gmx.de>

	markdowned the documentation and moved more of it out of the source file so the source code is shorter and it is clearer where to find what information

2010-05-13  Martin Haecker  <spamfaenger@gmx.de>

	added some documentation and moved a test to a more logical location

2010-05-13  Martin Haecker  <spamfaenger@gmx.de>

	updated changelog before release

2010-05-13  Martin Haecker  <spamfaenger@gmx.de>

	simplified test suite by introducing edit openEditor and enableEditor helpers that set some default default settings

2010-05-13  Martin Haecker  <spamfaenger@gmx.de>

	reworked all callbacks to receive the editor dom node as their this argument

2010-05-13  Martin Haecker  <spamfaenger@gmx.de>

	moved documentation into RELEASE NOTES and TODO files

2010-05-13  Martin Haecker  <spamfaenger@gmx.de>

	updated changelog before release

2010-05-13  Martin Haecker  <spamfaenger@gmx.de>

	the editor will now throw if you provide neither a callback or a url parameter to it. This lead to quite some fallout in the testsuite as it still enabled lots of editors with neither option...

2010-05-13  Martin Haecker  <spamfaenger@gmx.de>

	updated to jspec 4.3.1

2010-05-13  Martin Haecker  <spamfaenger@gmx.de>

	updated documentation to be easier to grasp and more in line to what the documentation of the homepage has been

2010-05-11  Martin Haecker  <spamfaenger@gmx.de>

	updated version to something more sane for the next release

2010-05-11  Martin Haecker  <spamfaenger@gmx.de>

	more documentation fixes and some generall small cleanup

2010-05-11  Martin Haecker  <spamfaenger@gmx.de>

	small enhancements to documentation but mostly removed and reordered TODOs and REFACTs

2010-05-11  Martin Haecker  <spamfaenger@gmx.de>

	improved documentation with an overview of how the inline editor works

2010-05-11  Martin Haecker  <spamfaenger@gmx.de>

	fixed missing var statements

2010-05-11  Martin Haecker  <spamfaenger@gmx.de>

	applied patch from robert that adds a cancel option like in jquery-ui to not open the editor when the click event happened on a cancel event

2010-05-10  Martin Haecker  <spamfaenger@gmx.de>

	added missing documentation

2010-05-10  Martin Haecker  <spamfaenger@gmx.de>

	enhanced callback protocol, added postclose and now checking the returnvalue of preinit. If it returns false, the editor is not opened

2010-05-07  Martin Haecker  <spamfaenger@gmx.de>

	fixed postclose callback to always be called after the editor has restored the new value either from the server or from the callback

2010-05-07  Martin Haecker  <spamfaenger@gmx.de>

	added postclose callback to do cleanup work after the editor has closed

2010-05-07  Martin Haecker  <spamfaenger@gmx.de>

	added preinit callback to the options

2010-05-06  Martin Haecker  <spamfaenger@gmx.de>

	added test and fix that return commits in all browsers

2010-05-04  Martin Haecker  <spamfaenger@gmx.de>

	removed stray debugger statement

2010-04-27  Martin Haecker  <spamfaenger@gmx.de>

	fixed test in firefox (background-color was not returned as part of background and transparent was used as the default background color for p elements (instead of inherit as we expected)) and added another fix for firefox to send blur events to other open editors.

2010-04-22  Martin Haecker  <spamfaenger@gmx.de>

	added documentation on how to use the callback interface for the callback submit

2010-04-22  Martin Haecker  <spamfaenger@gmx.de>

	added tests to verify that the animation color can be specified by a parameter

2010-04-22  Martin Haecker  <spamfaenger@gmx.de>

	added jquery ui

2010-04-22  Martin Haecker  <spamfaenger@gmx.de>

	added saving animation support

2010-04-22  Martin Haecker  <spamfaenger@gmx.de>

	fixed failuresOnly by not overwriting this.options in the tests and some small cleanups

2010-04-21  Martin Haecker  <spamfaenger@gmx.de>

	added a callback interface to submit to callback that allows the editor to know when the callback is done saving it's values

2010-04-21  Martin Haecker  <spamfaenger@gmx.de>

	added field type to tests that where applied to text, textarea and select fields

2010-04-21  Martin Haecker  <spamfaenger@gmx.de>

	added todos

2010-04-21  Martin Haecker  <spamfaenger@gmx.de>

	updated to 4.2.1 of jspec

2010-03-28  Martin Haecker  <spamfaenger@gmx.de>

	updated jspec to 4.1.0

2010-03-28  Martin Haecker  <spamfaenger@gmx.de>


spec/node.js

process.mixin(require('sys'))
require.paths.unshift('spec', './spec/lib', 'lib')
require('jspec')
require('unit/spec.helper')
require('yourlib')

quit = process.exit
print = puts

readFile = function(path) {
  var result
  require('posix')
    .cat(path, "utf8")
    .addCallback(function(contents){ result = contents })
    .addErrback(function(){ throw new Error("failed to read file `" + path + "'") })
    .wait()
  return result
}

JSpec
  .exec('spec/unit/spec.js')
  .run({ reporter: JSpec.reporters.Terminal, fixturePath: 'spec/fixtures' })
  .report()


spec/rhino.js

// Latest env.js doesn't work with jspec (yet)
load('./spec/support/env.rhino.js')
Envjs('spec/support/dummy.html')
load('./spec/support/jquery.js')

load('./spec/lib/jspec.js')
load('./spec/lib/jspec.xhr.js')


load('./spec/lib/jspec.jquery.js')

load('./spec/lib/jspec.growl.js')

load('lib/jquery.editinplace.js')
load('spec/unit/spec.helper.js')

JSpec
.exec('spec/unit/spec.js')
.run({ reporter: JSpec.reporters.Terminal, fixturePath: 'spec/fixtures' })
.report()

demo/js/demo.js
/*
 * Another In Place Editor - a jQuery edit in place plugin
 *
 * Copyright (c) 2009 Dave Hauenstein
 *
 * License:
 * This source file is subject to the BSD license bundled with this package.
 * Available online: {@link http://www.opensource.org/licenses/bsd-license.php}
 * If you did not receive a copy of the license, and are unable to obtain it,
 * email davehauenstein@gmail.com,
 * and I will send you a copy.
 *
 * Project home:
 * http://code.google.com/p/jquery-in-place-editor/
 *
 */
$(document).ready(function(){
	
	// All examples use the commit to function interface for ease of demonstration.
	// If you want to try it against a server, just comment the callback: and 
	// uncomment the url: lines.
	
	// The most basic form of using the inPlaceEditor
	$("#editme1").editInPlace({
		callback: function(unused, enteredText) { return enteredText; },
		// url: './server.php',
		show_buttons: true
	});

	// This example shows how to call the function and display a textarea
	// instead of a regular text box. A few other options are set as well,
	// including an image saving icon, rows and columns for the textarea,
	// and a different rollover color.
	$("#editme2").editInPlace({
		callback: function(unused, enteredText) { return enteredText; },
		// url: "./server.php",
		bg_over: "#cff",
		field_type: "textarea",
		textarea_rows: "15",
		textarea_cols: "35",
		saving_image: "./images/ajax-loader.gif"
	});

	// A select input field so we can limit our options
	$("#editme3").editInPlace({
		callback: function(unused, enteredText) { return enteredText; },
		// url: "./server.php",
		field_type: "select",
		select_options: "Change me to this, No way:no"
	});

	// Using a callback function to update 2 divs
	$("#editme4").editInPlace({
		callback: function(original_element, html, original){
			$("#updateDiv1").html("The original html was: " + original);
			$("#updateDiv2").html("The updated text is: " + html);
			return(html);
		}
	});
	
	$("#editme5").editInPlace({
		saving_animation_color: "#ECF2F8",
		callback: function(idOfEditor, enteredText, orinalHTMLContent, settingsParams, animationCallbacks) {
			animationCallbacks.didStartSaving();
			setTimeout(animationCallbacks.didEndSaving, 2000);
			return enteredText;
		}
	});
});

scripts/build_and_upload_release.sh
#!/bin/sh

# TODO: consider to create / update the tag when running the release script
# claim commits for my other email address

USAGE="Usage: $0 <bzr-tag-to-release> <googlecode-username> <googlecode-password>"

RELEASE_TAG=${1:? $USAGE}
USERNAME=${2:? $USAGE}
PASSWORD=${3:? $USAGE}

cd "$(dirname "$0")/.."

echo "Updating changelog"
bzr log --log-format=gnu-changelog > CHANGELOG
bzr commit CHANGELOG -m "updated changelog before release"

echo "Tagging release"
bzr tag "$RELEASE_TAG"

echo "Pushing all changes to upstream (googlecode)"
bzr push bm:upstream

echo "Building release zip"
mkdir -p build
EXPORTED_FILE="build/jquery-editInPlace-v${RELEASE_TAG}.zip"
bzr export --revision="$RELEASE_TAG" "$EXPORTED_FILE"

echo "Uploading release zip"
SUMMARY="jQuery In Place Editor v${RELEASE_TAG}"
python scripts/googlecode_upload.py \
	--summary="$SUMMARY" \
	--project="jquery-in-place-editor" \
	--user="$USERNAME" \
	--password="$PASSWORD" \
	--labels="Type-Archive,OpSys-All" \
	"$EXPORTED_FILE"

echo
echo
echo "Now you can feature this download at http://code.google.com/p/jquery-in-place-editor/downloads/list"
echo
echo

lib/jquery.editinplace.js
/*

A jQuery edit in place plugin

Version 2.2.0

Authors:
	Dave Hauenstein
	Martin Häcker <spamfaenger [at] gmx [dot] de>

Project home:
	http://code.google.com/p/jquery-in-place-editor/

Patches with tests welcomed! For guidance see the tests at </spec/unit/spec.js>. To submit, attach them to the bug tracker.

License:
This source file is subject to the BSD license bundled with this package.
Available online: {@link http://www.opensource.org/licenses/bsd-license.php}
If you did not receive a copy of the license, and are unable to obtain it, 
learn to use a search engine.

*/

(function($){

$.fn.editInPlace = function(options) {
	
	var settings = $.extend({}, $.fn.editInPlace.defaults, options);
	
	assertMandatorySettingsArePresent(settings);
	
	preloadImage(settings.saving_image);
	
	return this.each(function() {
		var dom = $(this);
		// This won't work with live queries as there is no specific element to attach this
		// one way to deal with this could be to store a reference to self and then compare that in click?
		if (dom.data('editInPlace'))
			return; // already an editor here
		dom.data('editInPlace', true);
		
		new InlineEditor(settings, dom).init();
	});
};

/// Switch these through the dictionary argument to $(aSelector).editInPlace(overideOptions)
/// Required Options: Either url or callback, so the editor knows what to do with the edited values.
$.fn.editInPlace.defaults = {
	url:				"", // string: POST URL to send edited content
	bg_over:			"#ffc", // string: background color of hover of unactivated editor
	bg_out:				"transparent", // string: background color on restore from hover
	hover_class:		"",  // string: class added to root element during hover. Will override bg_over and bg_out
	show_buttons:		false, // boolean: will show the buttons: cancel or save; will automatically cancel out the onBlur functionality
	save_button:		'<button class="inplace_save">Save</button>', // string: image button tag to use as “Save” button
	cancel_button:		'<button class="inplace_cancel">Cancel</button>', // string: image button tag to use as “Cancel” button
	params:				"", // string: example: first_name=dave&last_name=hauenstein extra paramters sent via the post request to the server
	field_type:			"text", // string: "text", "textarea", or "select";  The type of form field that will appear on instantiation
	default_text:		"(Click here to add text)", // string: text to show up if the element that has this functionality is empty
	use_html:			false, // boolean, set to true if the editor should use jQuery.fn.html() to extract the value to show from the dom node
	textarea_rows:		10, // integer: set rows attribute of textarea, if field_type is set to textarea
	textarea_cols:		25, // integer: set cols attribute of textarea, if field_type is set to textarea
	select_text:		"Choose new value", // string: default text to show up in select box
	select_options:		"", // string or array: Used if field_type is set to 'select'. Can be comma delimited list of options 'textandValue,text:value', Array of options ['textAndValue', 'text:value'] or array of arrays ['textAndValue', ['text', 'value']]. The last form is especially usefull if your labels or values contain colons)
	saving_text:		"Saving...", // string: text to be used when server is saving information
	saving_image:		"", // string: uses saving text specify an image location instead of text while server is saving
	saving_animation_color: 'transparent', // hex color string, will be the color the pulsing animation during the save pulses to. Note: Only works if jquery-ui is loaded
	value_required:		false, // boolean: if set to true, the element will not be saved unless a value is entered
	element_id:			"element_id", // string: name of parameter holding the id or the editable
	update_value:		"update_value", // string: name of parameter holding the updated/edited value
	original_value:		'original_value', // string: name of parameter holding the updated/edited value
	original_html:		"original_html", // string: name of parameter holding original_html value of the editable /* DEPRECATED in 2.2.0 */ use original_value instead.
	save_if_nothing_changed:	false,  // boolean: submit to function or server even if the user did not change anything
	on_blur:			"save", // string: "save" or null; what to do on blur; will be overridden if show_buttons is true
	cancel:				"", // string: if not empty, a jquery selector for elements that will not cause the editor to open even though they are clicked. E.g. if you have extra buttons inside editable fields
	
	// All callbacks will have this set to the DOM node of the editor that triggered the callback
	
	callback:			null, // function: function to be called when editing is complete; cancels ajax submission to the url param. Prototype: function(idOfEditor, enteredText, orinalHTMLContent, settingsParams, callbacks). The function needs to return the value that should be shown in the dom. Returning undefined means cancel and will restore the dom and trigger an error. callbacks is a dictionary with two functions didStartSaving and didEndSaving() that you can use to tell the inline editor that it should start and stop any saving animations it has configured. /* DEPRECATED in 2.1.0 */ Parameter idOfEditor, use $(this).attr('id') instead
	success:			null, // function: this function gets called if server responds with a success. Prototype: function(newEditorContentString)
	error:				null, // function: this function gets called if server responds with an error. Prototype: function(request)
	error_sink:			function(idOfEditor, errorString) { alert(errorString); }, // function: gets id of the editor and the error. Make sure the editor has an id, or it will just be undefined. If set to null, no error will be reported. /* DEPRECATED in 2.1.0 */ Parameter idOfEditor, use $(this).attr('id') instead
	preinit:			null, // function: this function gets called after a click on an editable element but before the editor opens. If you return false, the inline editor will not open. Prototype: function(currentDomNode). /* DEPRECATED in 2.1.0 */ Parameter currentDomNode, use $(this) instead.
	postclose:			null // function: this function gets called after the inline editor has closed and all values are updated. Prototype: function(currentDomNode). /* DEPRECATED in 2.1.0 */ Parameter currentDomNode, use $(this) instead.
};


function InlineEditor(settings, dom) {
	this.settings = settings;
	this.dom = dom;
	this.originalValue = null;
	this.didInsertDefaultText = false;
	this.shouldDelayReinit = false;
};

$.extend(InlineEditor.prototype, {
	
	init: function() {
		this.setDefaultTextIfNeccessary();
		this.connectEvents();
	},
	
	reinit: function() {
		if (this.shouldDelayReinit)
			return;
		
		if (this.settings.postclose)
			this.triggerCallback(this.settings.postclose, /* DEPRECATED in 2.1.0 */ this.dom);
		
		this.markEditorAsInactive();
		this.connectEvents();
	},
	
	setDefaultTextIfNeccessary: function() {
		if('' !== this.dom.html())
			return;
		
		this.dom.html(this.settings.default_text);
		this.didInsertDefaultText = true;
	},
	
	connectEvents: function() {
		var that = this;
		this.dom
			.bind('mouseenter.editInPlace', function(){ that.addHoverEffect(); })
			.bind('mouseleave.editInPlace', function(){ that.removeHoverEffect(); })
			.bind('click.editInPlace', function(anEvent){ that.openEditor(anEvent); });
	},
	
	addHoverEffect: function() {
		if (this.settings.hover_class)
			this.dom.addClass(this.settings.hover_class);
		else
			this.dom.css("background-color", this.settings.bg_over);
	},
	
	removeHoverEffect: function() {
		if (this.settings.hover_class)
			this.dom.removeClass(this.settings.hover_class);
		else
			this.dom.css("background-color", this.settings.bg_out);
	},
	
	openEditor: function(anEvent) {
		if ( ! this.shouldOpenEditor(anEvent.target))
			return;
		
		// prevent re-opening the editor when it is already open
		this.dom.unbind('.editInPlace');
		this.removeHoverEffect();
		this.removeInsertedDefaultTextIfNeccessary();
		this.saveOriginalValue();
		this.workaroundFirefoxBlurBug();
		this.markEditorAsActive();
		this.replaceContentWithEditor();
		this.connectEventsToEditor();
	},
	
	shouldOpenEditor: function(eventTarget) {
		if (this.isClickedObjectCancelled(eventTarget))
			return false;
		
		if (this.settings.preinit)
			return false !== this.triggerCallback(this.settings.preinit, /* DEPRECATED in 2.1.0 */ this.dom);
		
		return true;
	},
	
	removeInsertedDefaultTextIfNeccessary: function() {
		if ( ! this.didInsertDefaultText
			|| this.dom.html() !== this.settings.default_text)
			return;
		
		this.dom.html('');
		this.didInsertDefaultText = false;
	},
	
	isClickedObjectCancelled: function(eventTarget) {
		if ( ! this.settings.cancel)
			return false;
		
		var eventTargetAndParents = $(eventTarget).parents().andSelf();
		var elementsMatchingCancelSelector = eventTargetAndParents.filter(this.settings.cancel);
		return 0 !== elementsMatchingCancelSelector.length;
	},
	
	saveOriginalValue: function() {
		if (this.settings.use_html)
			this.originalValue = this.dom.html();
		else
			this.originalValue = trim(this.dom.text());
	},
	
	restoreOriginalValue: function() {
		if (this.settings.use_html)
			this.dom.html(this.originalValue);
		else
			this.dom.text(this.originalValue);
	},
	
	workaroundFirefoxBlurBug: function() {
		if ( ! $.browser.mozilla)
			return;
		
		// TODO: Opera seems to also have this bug....
		
		// Firefox will forget to send a blur event to an input element when another one is
		// created and selected programmatically. This means that if another inline editor is
		// opened, existing inline editors will _not_ close if they are configured to submit when blurred.
		// This is actually the first time I've written browser specific code for a browser different than IE! Wohoo!
		
		// Using parents() instead document as base to workaround the fact that in the unittests
		// the editor is not a child of window.document but of a document fragment
		this.dom.parents(':last').find('.editInPlace-active :input').blur();
	},
	
	replaceContentWithEditor: function() {
		var buttons_html  = (this.settings.show_buttons) ? this.settings.save_button + ' ' + this.settings.cancel_button : '';
		var editorElement = this.createEditorElement(); // needs to happen before anything is replaced
		/* insert the new in place form after the element they click, then empty out the original element */
		this.dom.html('<form class="inplace_form" style="display: inline; margin: 0; padding: 0;"></form>')
			.find('form')
				.append(editorElement)
				.append(buttons_html);
	},
	
	createEditorElement: function() {
		if (-1 === $.inArray(this.settings.field_type, ['text', 'textarea', 'select']))
			throw "Unknown field_type <fnord>, supported are 'text', 'textarea' and 'select'";
		
		if ("select" === this.settings.field_type)
			return this.createSelectEditor();
		
		var editor = null;
		if ("text" === this.settings.field_type)
			editor = $('<input type="text"' + this.inputNameAndClass() + '/>');
		else if ("textarea" === this.settings.field_type)
			editor = $('<textarea' + this.inputNameAndClass() 
				+ 'rows="' + this.settings.textarea_rows + '" cols="' 
				+ this.settings.textarea_cols + '"></textarea>');
		
		editor.val(this.originalValue);
		return editor;
	},
	
	inputNameAndClass: function() {
		return ' name="inplace_value" class="inplace_field" ';
	},
	
	createSelectEditor: function() {
		var editor = $('<select' + this.inputNameAndClass() + '>'
			+	'<option disabled="true" value="">' + this.settings.select_text + '</option>'
			+ '</select>');
		
		var optionsArray = this.settings.select_options;
		if ( ! $.isArray(optionsArray))
			optionsArray = optionsArray.split(',');
			
		for (var i=0; i<optionsArray.length; i++) {
			
			var currentTextAndValue = optionsArray[i];
			if ( ! $.isArray(currentTextAndValue))
				currentTextAndValue = currentTextAndValue.split(':');
			
			var value = trim(currentTextAndValue[1] || currentTextAndValue[0]);
			var text = trim(currentTextAndValue[0]);
			
			var selected = (value == this.originalValue) ? 'selected="selected" ' : '';
			var option = $('<option ' + selected + ' ></option>').val(value).text(text);
			editor.append(option);
		}
		return editor;
		
	},
	
	connectEventsToEditor: function() {
		var that = this;
		function cancelEditorAction() {
			that.handleCancelEditor();
			return false; // stop event bubbling
		}
		function saveEditorAction() {
			that.handleSaveEditor();
			return false; // stop event bubbling
		}
		
		var form = this.dom.find("form");
		
		form.find(".inplace_field").focus().select();
		form.find(".inplace_cancel").click(cancelEditorAction);
		form.find(".inplace_save").click(saveEditorAction);
		
		if ( ! this.settings.show_buttons) {
				// TODO: Firefox has a bug where blur is not reliably called when focus is lost 
				//       (for example by another editor appearing)
			if ("save" === this.settings.on_blur)
				form.find(".inplace_field").blur(saveEditorAction);
			else
				form.find(".inplace_field").blur(cancelEditorAction);
			
			// workaround for firefox bug where it won't submit on enter if no button is shown
			if ($.browser.mozilla) {
				form.keyup(function(event) {
					if (13 === event.which)
						saveEditorAction();
				});
			}
		}
		
		// allow canceling with escape
		form.keyup(function(event) {
			if (27 === event.which) // escape
				return cancelEditorAction();
			else if (13 === event.which) // return (webkit and chrome nightlies don't commit otherwise?
				return saveEditorAction();
		});
		
		
		form.submit(saveEditorAction);
	},
	
	handleCancelEditor: function() {
		this.restoreOriginalValue();
		this.reinit();
	},
	
	handleSaveEditor: function() {
		var enteredText = this.dom.find(':input').val();
		
		if (this.isDisabledDefaultSelectChoice()
			|| this.isUnchangedInput(enteredText)) {
			this.handleCancelEditor();
			return;
		}
		
		if (this.didForgetRequiredText(enteredText)) {
			this.handleCancelEditor();
			this.reportError("Error: You must enter a value to save this field");
			return;
		}
		
		this.showSaving();
		
		if (this.settings.callback)
			this.handleSubmitToCallback(enteredText);
		else
			this.handleSubmitToServer(enteredText);
	},
	
	didForgetRequiredText: function(enteredText) {
		return this.settings.value_required 
			&& ("" === enteredText || undefined === enteredText);
	},
	
	isDisabledDefaultSelectChoice: function() {
		return this.dom.find('option').eq(0).is(':selected:disabled');
	},
	
	isUnchangedInput: function(enteredText) {
		return ! this.settings.save_if_nothing_changed
			&& this.originalValue === enteredText;
	},
	
	showSaving: function() {
		var saving_message = this.settings.saving_text;
		if("" !== this.settings.saving_image)
			// REFACT: alt should be the configured saving message
			saving_message = '<img src="' + this.settings.saving_image + '" alt="Saving..." />';
		this.dom.html(saving_message);
	},
	
	handleSubmitToCallback: function(enteredText) {
		// REFACT: consider to encode enteredText and originalHTML before giving it to the callback
		this.enableOrDisableAnimationCallbacks(true, false);
		var newHTML = this.triggerCallback(this.settings.callback, /* DEPRECATED in 2.1.0 */ this.id(), enteredText, this.originalValue, 
			this.settings.params, this.savingAnimationCallbacks());
		
		if (undefined === newHTML) {
			/* failure; put original back */
			this.reportError("Error: Failed to save value: " + enteredText);
			this.restoreOriginalValue();
		}
		else
			this.dom.html(newHTML);
		
		if (this.didCallNoCallbacks()) {
			this.enableOrDisableAnimationCallbacks(false, false);
			this.reinit();
		}
	},
	
	handleSubmitToServer: function(enteredText) {
		var data = this.settings.update_value + '=' + encodeURIComponent(enteredText) 
			+ '&' + this.settings.element_id + '=' + this.dom.attr("id") 
			+ ((this.settings.params) ? '&' + this.settings.params : '')
			+ '&' + this.settings.original_html + '=' + encodeURIComponent(this.originalValue) /* DEPRECATED in 2.2.0 */
			+ '&' + this.settings.original_value + '=' + encodeURIComponent(this.originalValue);
		
		this.enableOrDisableAnimationCallbacks(true, false);
		this.didStartSaving();
		var that = this;
		$.ajax({
			url: that.settings.url,
			type: "POST",
			data: data,
			dataType: "html",
			complete: function(request){
				that.didEndSaving();
			},
			success: function(html){
				var new_text = html || that.settings.default_text;
				
				/* put the newly updated info into the original element */
				that.dom.html(new_text);
				// REFACT: remove dom parameter, already in this, not documented, should be easy to remove
				// REFACT: callback should be able to override what gets put into the DOM
				that.triggerCallback(that.settings.success, html);
			},
			error: function(request) {
				that.dom.html(that.originalHTML); // REFACT: what about a restorePreEditingContent()
				if (that.settings.error)
					// REFACT: remove dom parameter, already in this, not documented, can remove without deprecation
					// REFACT: callback should be able to override what gets entered into the DOM
					that.triggerCallback(that.settings.error, request);
				else
					that.reportError("Failed to save value: " + request.responseText || 'Unspecified Error');
			}
		});
	},
	
	// Utilities .........................................................
	
	triggerCallback: function(aCallback /*, arguments */) {
		if ( ! aCallback)
			return; // callback wasn't specified after all
		
		var callbackArguments = Array.prototype.splice.call(arguments, 1);
		return aCallback.apply(this.dom[0], callbackArguments);
	},
	
	reportError: function(anErrorString) {
		this.triggerCallback(this.settings.error_sink, /* DEPRECATED in 2.1.0 */ this.id(), anErrorString);
	},
	
	// REFACT: this method should go, callbacks should get the dom node itself as an argument
	id: function() {
		return this.dom.attr('id');
	},
	
	markEditorAsActive: function() {
		this.dom.addClass('editInPlace-active');
	},
	
	markEditorAsInactive: function() {
		this.dom.removeClass('editInPlace-active');
	},
	
	// REFACT: consider rename, doesn't deal with animation directly
	savingAnimationCallbacks: function() {
		var that = this;
		return {
			didStartSaving: function() { that.didStartSaving(); },
			didEndSaving: function() { that.didEndSaving(); }
		};
	},
	
	enableOrDisableAnimationCallbacks: function(shouldEnableStart, shouldEnableEnd) {
		this.didStartSaving.enabled = shouldEnableStart;
		this.didEndSaving.enabled = shouldEnableEnd;
	},
	
	didCallNoCallbacks: function() {
		return this.didStartSaving.enabled && ! this.didEndSaving.enabled;
	},
	
	assertCanCall: function(methodName) {
		if ( ! this[methodName].enabled)
			throw new Error('Cannot call ' + methodName + ' now. See documentation for details.');
	},
	
	didStartSaving: function() {
		this.assertCanCall('didStartSaving');
		this.shouldDelayReinit = true;
		this.enableOrDisableAnimationCallbacks(false, true);
		
		this.startSavingAnimation();
	},
	
	didEndSaving: function() {
		this.assertCanCall('didEndSaving');
		this.shouldDelayReinit = false;
		this.enableOrDisableAnimationCallbacks(false, false);
		this.reinit();
		
		this.stopSavingAnimation();
	},
	
	startSavingAnimation: function() {
		var that = this;
		this.dom
			.animate({ backgroundColor: this.settings.saving_animation_color }, 400)
			.animate({ backgroundColor: 'transparent'}, 400, 'swing', function(){
				// In the tests animations are turned off - i.e they happen instantaneously.
				// Hence we need to prevent this from becomming an unbounded recursion.
				setTimeout(function(){ that.startSavingAnimation(); }, 10);
			});
	},
	
	stopSavingAnimation: function() {
		this.dom
			.stop(true)
			.css({backgroundColor: ''});
	},
	
	missingCommaErrorPreventer:''
});



// Private helpers .......................................................

function assertMandatorySettingsArePresent(options) {
	// one of these needs to be non falsy
	if (options.url || options.callback)
		return;
	
	throw new Error("Need to set either url: or callback: option for the inline editor to work.");
}

/* preload the loading icon if it is configured */
function preloadImage(anImageURL) {
	if ('' === anImageURL)
		return;
	
	var loading_image = new Image();
	loading_image.src = anImageURL;
}

function trim(aString) {
	return aString
		.replace(/^\s+/, '')
		.replace(/\s+$/, '');
}

})(jQuery);
